engine/src/main/java/io/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java

```diff
@@ -71,6 +73,19 @@ public void applyState(final long elementInstanceKey, final ProcessInstanceRecor
         variableState.removeTemporaryVariables(flowScopeInstance.getKey());
       }
     }
+
+    // manage the multi-instance loop counter
+    if (flowScopeElementType == BpmnElementType.MULTI_INSTANCE_BODY
+        && MigratedStreamProcessors.isMigrated(currentElementType)) {
```

Comment:
:information_source: We should not apply this logic for non-migrated processes. Otherwise, we increment the loop counter too often.